File: frontend/lib/screens/dashboard/schedule/models/schedule_item.dart
class ScheduleItem {
  final String? id;
  final String userId;
  final String type;
  final String title;

  final String? day;               // Only for Classes
  final String? startTime;         // String "3:30 PM"
  final String? endTime;           // String "4:30 PM"

  final String? room;
  final String? details;

  final DateTime? startDate;       // quizzes/assignments/others
  final DateTime? endDate;         // optional

  ScheduleItem({
    this.id,
    required this.userId,
    required this.type,
    required this.title,
    this.day,
    this.startTime,
    this.endTime,
    this.room,
    this.details,
    this.startDate,
    this.endDate,
  });

  factory ScheduleItem.fromJson(Map<String, dynamic> json) => ScheduleItem(
    id: json['id'],
    userId: json['userId'],
    type: json['type'],
    title: json['title'],
    day: json['day'],
    startTime: json['startTime'],
    endTime: json['endTime'],
    room: json['room'],
    details: json['details'],
    startDate: json['startDate'] != null ? DateTime.parse(json['startDate']) : null,
    endDate: json['endDate'] != null ? DateTime.parse(json['endDate']) : null,
  );

  Map<String, dynamic> toJson() => {
    if (id != null) "id": id,
    "userId": userId,
    "type": type,
    "title": title,
    if (day != null) "day": day,
    if (startTime != null) "startTime": startTime,
    if (endTime != null) "endTime": endTime,
    if (room != null) "room": room,
    if (details != null) "details": details,
    if (startDate != null) "startDate": startDate!.toIso8601String(),
    if (endDate != null) "endDate": endDate!.toIso8601String(),
  };
}

File: backend/models/ScheduleItem.js
import mongoose from "mongoose";

const scheduleItemSchema = new mongoose.Schema({
  userId: { type: String, required: true },
  type: { type: String, enum: ["Classes","Quizzes","Assignments","Others"], required: true },
  title: { type: String, required: true },
   day: { type: String },
  startDate: { type: Date },      
  endDate: { type: Date },        
  startTime: { type: String },    
  endTime: { type: String },      
  room: { type: String },
  details: { type: String },
  createdAt: { type: Date, default: Date.now }
});

// Return id instead of _id
scheduleItemSchema.set('toJSON', {
  virtuals: true,
  versionKey: false,
  transform: (doc, ret) => { delete ret._id; }
});


scheduleItemSchema.pre("save", function (next) {
  // Only validate if startDate and endDate are provided
  if (this.startDate && this.endDate) {
    if (this.endDate < this.startDate) {
      return next(new Error("End date must be after start date."));
    }
  }

  // Only validate time if startTime and endTime are provided
  if (this.startTime && this.endTime) {
    // Parse time strings into minutes
    const parseTime = (t) => {
      const [hourMin, period] = t.split(" ");
      let [h, m] = hourMin.split(":").map(Number);
      if (period.toUpperCase() === "PM" && h !== 12) h += 12;
      if (period.toUpperCase() === "AM" && h === 12) h = 0;
      return h * 60 + m;
    };

    const startMinutes = parseTime(this.startTime);
    const endMinutes = parseTime(this.endTime);

    // Only enforce endTime > startTime if startDate === endDate or endDate is missing
    const startDateTime = this.startDate ? new Date(this.startDate) : null;
    const endDateTime = this.endDate ? new Date(this.endDate) : null;

    if (
      (!startDateTime && !endDateTime) || // both dates missing
      (startDateTime && endDateTime && startDateTime.getTime() === endDateTime.getTime())
    ) {
      if (endMinutes <= startMinutes) {
        return next(new Error("End time must be after start time for the same day."));
      }
    }
  }

  next();
});


const ScheduleItem = mongoose.model("ScheduleItem", scheduleItemSchema);
export default ScheduleItem;
